High latency with multiple drones

We are trying to perform an experiment with the crazyflie drones. For our experiment we are aiming to use 12 drones. However we have been having trouble with running multiple drones at the same time. Our setup consists of the 

- crazyflie 2.1 drones (https://www.bitcraze.io/products/old-products/crazyflie-2-1/)
- 4 crazyradio 2.0 (https://www.bitcraze.io/products/crazyradio-2-0/)
- qualisys motion capture system

We are using identical marker arrangements with 3 markers for tracking

In our testing we have been using 6 crazyflies with 2 crazyradios, 3 drones assigned to each. We have had troubles with getting all 6 drones to run succesfully. We are running the hello_all script provided in the crazyswarm library. Very often only 4 or 5 of the drones will take off. The other drones will flip immediatly. When all drones do take off succesfully several drones will look highly unstable causing them to fall out of the sky.

We have attempted to fix all hardware issues. We confirmed this by running all the crazyflie drones by themselves. When running the drones by themselves none of them showed any significant problems and they showed very stable behaviour. 

we have tried several different frequencies for the communication with the motion capture system (50-200Hz), however there wasn't much of a visible difference between any of these rates. Additionaly we have attempted to assign different radio channels to the crazyflies such that each radio works on its own channel to minimize possible interference. This also didn't have any visible effect.

We do notice there is a lot of latency, even on a singular drones as can be seen here:

```bash
ros2 launch crazyflie launch.py
[INFO] [launch]: All log files can be found below /home/_/.ros/log/2025-11-06-18-51-38-352732-dhsg-supercomputer-2895057
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [motion_capture_tracking_node-1]: process started with pid [2895073]
[INFO] [crazyflie_server-2]: process started with pid [2895075]
[INFO] [teleop-3]: process started with pid [2895077]
[INFO] [joy_node-4]: process started with pid [2895079]
[motion_capture_tracking_node-1] logClouds=0
[teleop-3] [INFO] [1762451498.422021410] [teleop]: Mode changed to high_level
[crazyflie_server-2] [WARN] [1762451498.473371170] [crazyflie_server]: [cf11] Received unrequested data for block: 1
[crazyflie_server-2] [WARN] [1762451498.490491405] [crazyflie_server]: [cf11] Received unrequested data for block: 1
[crazyflie_server-2] [INFO] [1762451498.492898413] [crazyflie_server]: [cf11] Requesting parameters...
[crazyflie_server-2] [INFO] [1762451498.510186481] [crazyflie_server]: [cf11] Param TOC: 279 entries with CRC 51380503
[crazyflie_server-2] [INFO] [1762451498.510337272] [crazyflie_server]: [cf11] Param TOC: found cache.
[crazyflie_server-2] [INFO] [1762451498.527346801] [crazyflie_server]: [cf11] reqParamTOC: 0.101988 s (279 params)
[crazyflie_server-2] [INFO] [1762451498.527411158] [crazyflie_server]: [cf11] Update parameter "commander.enHighLevel" to 1
[crazyflie_server-2] [INFO] [1762451498.527433058] [crazyflie_server]: [cf11] Update parameter "locSrv.extPosStdDev" to 0.001000
[crazyflie_server-2] [INFO] [1762451498.527446876] [crazyflie_server]: [cf11] Update parameter "locSrv.extQuatStdDev" to 0.050000
[crazyflie_server-2] [INFO] [1762451498.527458737] [crazyflie_server]: [cf11] Update parameter "stabilizer.controller" to 2
[crazyflie_server-2] [INFO] [1762451498.527470081] [crazyflie_server]: [cf11] Update parameter "stabilizer.estimator" to 2
[crazyflie_server-2] [INFO] [1762451498.527485521] [crazyflie_server]: [cf11] ref frame: mocap
[crazyflie_server-2] [INFO] [1762451498.535108744] [crazyflie_server]: [cf11] Log TOC: 485 entries with CRC 3375487317
[crazyflie_server-2] [INFO] [1762451498.535185044] [crazyflie_server]: [cf11] Log TOC: found cache.
[crazyflie_server-2] [INFO] [1762451498.535418630] [crazyflie_server]: [cf11] Logging to low_level_control at 50 Hz
[crazyflie_server-2] [INFO] [1762451498.550437540] [crazyflie_server]: [cf11] Logging to /pose at 100 Hz
[crazyflie_server-2] [INFO] [1762451498.558676119] [crazyflie_server]: [cf11] Logging to /status at 1 Hz
[crazyflie_server-2] [WARN] [1762451498.559511200] [crazyflie_server]: [cf11] Older firmware. status/num_rx_broadcast and status/num_rx_unicast are set to zero.
[crazyflie_server-2] [INFO] [1762451498.572028816] [crazyflie_server]: [cf11] Requesting memories...
[crazyflie_server-2] [INFO] [1762451498.576904258] [crazyflie_server]: [cf11] Memories: 4
[crazyflie_server-2] [WARN] [1762451499.419750752] [crazyflie_server]: [all] Motion capture rate off (#: 6, Avg: 101.7, Min: 84.1, Max: 165.5)
[crazyflie_server-2] [WARN] [1762451499.425430773] [crazyflie_server]: [cf11] last latency update: 1.002006 s
[crazyflie_server-2] [WARN] [1762451500.419595059] [crazyflie_server]: [all] Motion capture rate off (#: 3, Avg: 100.9, Min: 82.3, Max: 123.3)
[crazyflie_server-2] [WARN] [1762451500.425397338] [crazyflie_server]: [cf11] last latency update: 2.001974 s
[crazyflie_server-2] [INFO] [1762451500.535402242] [crazyflie_server]: [cf11] SYS: ----------------------------
[crazyflie_server-2] [INFO] [1762451500.560358610] [crazyflie_server]: [cf11] SYS: Crazyflie 2.1 is up and running!
[crazyflie_server-2] [INFO] [1762451500.574036165] [crazyflie_server]: [cf11] SYS: Build 0:e527895e8c6e (2023.11) CLEAN
[crazyflie_server-2] [INFO] [1762451500.606354188] [crazyflie_server]: [cf11] SYS: I am 0x32303831353351100056002D and I have 1024KB of flash!
[crazyflie_server-2] [INFO] [1762451500.612363704] [crazyflie_server]: [cf11] CFGBLK: v1, verification [OK]
[crazyflie_server-2] [WARN] [1762451501.419653426] [crazyflie_server]: [all] Motion capture rate off (#: 5, Avg: 101.0, Min: 83.4, Max: 139.1)
[crazyflie_server-2] [WARN] [1762451501.425436978] [crazyflie_server]: [cf11] last latency update: 3.002012 s
[crazyflie_server-2] [WARN] [1762451501.436360457] [crazyflie_server]: [cf11] High latency: 10.9 ms
[crazyflie_server-2] [INFO] [1762451501.491351702] [crazyflie_server]: [cf11] IMU: BMI088 Gyro connection [OK].
[crazyflie_server-2] [INFO] [1762451501.541377306] [crazyflie_server]: [cf11] IMU: BMI088 Accel connection [OK]
[crazyflie_server-2] [INFO] [1762451501.560347570] [crazyflie_server]: [cf11] IMU: BMP388 I2C connection [OK]
[crazyflie_server-2] [INFO] [1762451501.582353805] [crazyflie_server]: [cf11] ESTIMATOR: Using Complementary (1) estimator
[crazyflie_server-2] [INFO] [1762451501.592361288] [crazyflie_server]: [cf11] CONTROLLER: Using PID (1) contriver
[crazyflie_server-2] [INFO] [1762451502.074358493] [crazyflie_server]: [cf11] EEPROM: I2C connection [OK].
[crazyflie_server-2] [INFO] [1762451502.077385091] [crazyflie_server]: [cf11] STORAGE: Storage check [OK].
[crazyflie_server-2] [INFO] [1762451502.086346346] [crazyflie_server]: [cf11] IMU: BMI088 gyro self-test [OK]
[crazyflie_server-2] [WARN] [1762451502.419600282] [crazyflie_server]: [all] Motion capture rate off (#: 5, Avg: 101.1, Min: 81.3, Max: 126.1)
[crazyflie_server-2] [INFO] [1762451502.877347262] [crazyflie_server]: [cf11] SYS: The system resumed after a failed assert [WARNING]
[crazyflie_server-2] [INFO] [1762451502.882369193] [crazyflie_server]: [cf11] SYS: Assert failed at ../src/hal/src/radiolink.c:164
[crazyflie_server-2] [INFO] [1762451502.885359927] [crazyflie_server]: [cf11] SYS: cfAssertNormalStart [FAIL]
[crazyflie_server-2] [WARN] [1762451503.419601513] [crazyflie_server]: [all] Motion capture rate off (#: 7, Avg: 100.8, Min: 82.8, Max: 127.2)
[crazyflie_server-2] [WARN] [1762451504.419606351] [crazyflie_server]: [all] Motion capture rate off (#: 4, Avg: 100.8, Min: 83.0, Max: 122.8)
[crazyflie_server-2] [WARN] [1762451504.436386692] [crazyflie_server]: [cf11] High latency: 11.0 ms
[crazyflie_server-2] [WARN] [1762451504.438371952] [crazyflie_server]: [cf11] High latency: 12.9 ms
[crazyflie_server-2] [WARN] [1762451504.439381356] [crazyflie_server]: [cf11] High latency: 14.0 ms
[crazyflie_server-2] [WARN] [1762451505.419607117] [crazyflie_server]: [all] Motion capture rate off (#: 1, Avg: 100.7, Min: 84.2, Max: 120.6)
[crazyflie_server-2] [WARN] [1762451505.436387549] [crazyflie_server]: [cf11] High latency: 10.9 ms
[crazyflie_server-2] [WARN] [1762451505.437410612] [crazyflie_server]: [cf11] High latency: 12.0 ms
[crazyflie_server-2] [WARN] [1762451505.439341029] [crazyflie_server]: [cf11] High latency: 13.9 ms
[crazyflie_server-2] [WARN] [1762451505.442329194] [crazyflie_server]: [cf11] High latency: 16.9 ms
[crazyflie_server-2] [WARN] [1762451506.419645294] [crazyflie_server]: [all] Motion capture rate off (#: 4, Avg: 101.0, Min: 77.9, Max: 143.3)
[crazyflie_server-2] [WARN] [1762451506.436389168] [crazyflie_server]: [cf11] High latency: 10.9 ms
[crazyflie_server-2] [WARN] [1762451507.419630438] [crazyflie_server]: [all] Motion capture rate off (#: 9, Avg: 101.4, Min: 80.3, Max: 127.9)
[crazyflie_server-2] [WARN] [1762451507.436373683] [crazyflie_server]: [cf11] High latency: 11.0 ms
[crazyflie_server-2] [WARN] [1762451507.437361864] [crazyflie_server]: [cf11] High latency: 11.9 ms
[crazyflie_server-2] [WARN] [1762451507.438409581] [crazyflie_server]: [cf11] High latency: 13.0 ms
[crazyflie_server-2] [WARN] [1762451507.441378066] [crazyflie_server]: [cf11] High latency: 16.0 ms
[crazyflie_server-2] [WARN] [1762451508.419636327] [crazyflie_server]: [all] Motion capture rate off (#: 6, Avg: 101.0, Min: 81.1, Max: 130.0)
[crazyflie_server-2] [WARN] [1762451509.419631546] [crazyflie_server]: [all] Motion capture rate off (#: 4, Avg: 100.8, Min: 79.8, Max: 136.5)
^C[WARNING] [launch]: user interrupted with ctrl-c (SIGINT)
[teleop-3] [INFO] [1762451509.865077946] [rclcpp]: signal_handler(signum=2)
[crazyflie_server-2] [INFO] [1762451509.864870601] [rclcpp]: signal_handler(signum=2)
[joy_node-4] [INFO] [1762451509.864870162] [rclcpp]: signal_handler(signum=2)
[motion_capture_tracking_node-1] [INFO] [1762451509.864870037] [rclcpp]: signal_handler(signum=2)
[motion_capture_tracking_node-1] terminate called after throwing an instance of 'rclcpp::exceptions::RCLError'
[motion_capture_tracking_node-1]   what():  failed to create guard condition: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at ./src/rcl/guard_condition.c:67
[ERROR] [motion_capture_tracking_node-1]: process has died [pid 2895073, exit code -6, cmd '/opt/ros/humble/lib/motion_capture_tracking/motion_capture_tracking_node --ros-args -r __node:=motion_capture_tracking --params-file /tmp/launch_params_dlp1jkrl'].
[INFO] [teleop-3]: process has finished cleanly [pid 2895077]
[INFO] [crazyflie_server-2]: process has finished cleanly [pid 2895075]
[INFO] [joy_node-4]: process has finished cleanly [pid 2895079]
```

--- a/src/legacy_usb.c
+++ b/src/legacy_usb.c
@@ -18,14 +18,41 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(usb);
 
-struct usb_command {
+enum command_type {
+    command_data,
+    command_setup,
+};
+
+char* command_type_name[] = {
+    "data",
+    "setup",
+};
+
+struct data_command {
     char payload[64];
     uint32_t length;
 };
 
+struct setup_command {
+    struct usb_setup_packet setup_packet;
+    uint32_t length;
+    char data[16];
+};
+
+struct usb_command {
+    enum command_type type;
+    union {
+        struct data_command data;
+        struct setup_command setup;
+    };
+};
+
 K_MSGQ_DEFINE(command_queue, sizeof(struct usb_command), 10, 4);
 
+K_MUTEX_DEFINE(usb_radio_mutex);
+
 static void fw_scan(uint8_t start, uint8_t stop, char* data, int data_length);
+static void handle_vendor_command(struct setup_command* setup);
 
 // state
 static struct {
@@ -86,14 +113,16 @@ void crazyradio_out_cb(uint8_t ep, enum usb_dc_ep_cb_status_code cb_status)
     uint32_t bytes_to_read;
     static struct usb_command command;
 
-	usb_read(ep, NULL, 0, &bytes_to_read);
+    command.type = command_data;
+	
+    usb_read(ep, NULL, 0, &bytes_to_read);
 	LOG_DBG("ep 0x%x, bytes to read %d ", ep, bytes_to_read);
     if (bytes_to_read > 64) {
         bytes_to_read = 64;
     }
-	usb_read(ep, command.payload, bytes_to_read, NULL);
+	usb_read(ep, command.data.payload, bytes_to_read, NULL);
 
-	command.length = bytes_to_read;
+	command.data.length = bytes_to_read;
 
     k_msgq_put(&command_queue, &command, K_FOREVER);
 
@@ -143,64 +172,58 @@ static int crazyradio_vendor_handler(struct usb_setup_packet *setup,
 {
 	LOG_DBG("Class request: bRequest 0x%x bmRequestType 0x%x len %d",
 		setup->bRequest, setup->bmRequestType, *len);
-	
+
+    static struct usb_command command;
+
 	if (USB_REQTYPE_GET_TYPE(setup->bmRequestType) == USB_REQTYPE_TYPE_VENDOR) {
-		
-		if (setup->bRequest == SET_RADIO_CHANNEL && setup->wLength == 0) {
-			LOG_DBG("Setting radio channel %d", setup->wValue);
-			uint16_t channel = setup->wValue;
-			if (channel <= 100) {
-				esb_set_channel(channel);
-			}
-			// If channel >100 used, packets will be ignored (ie. virtually not acked)
-			state.channel = channel;
-		} else if (setup->bRequest == SET_RADIO_ADDRESS && setup->wLength == 5) {
-			LOG_DBG("Setting radio address %02x%02x%02x%02x%02x", (unsigned int)(*data)[0], (unsigned int)(*data)[1], (unsigned int)(*data)[2], (unsigned int)(*data)[3], (unsigned int)(*data)[4]);
-			esb_set_address(*data);
-		} else if (setup->bRequest == SET_DATA_RATE && setup->wLength == 0 && setup->wValue < 3) {
-			char* datarates[] = {"250K", "1M", "2M"};
-			LOG_DBG("Setting radio datarate to %s", datarates[setup->wValue]);
-			uint32_t datarate = setup->wValue;
-            if (datarate == 1) {
-				esb_set_bitrate(radioBitrate1M);
-            } else if (datarate == 2) {
-                esb_set_bitrate(radioBitrate2M);
+        LOG_DBG("Vendor request: bRequest 0x%x bmRequestType 0x%x len %d",
+                setup->bRequest, setup->bmRequestType, *len);
+
+        if (setup->bRequest == SET_RADIO_CHANNEL ||
+            setup->bRequest == SET_RADIO_ADDRESS ||
+            setup->bRequest == SET_DATA_RATE ||
+            setup->bRequest == SET_RADIO_POWER ||
+            setup->bRequest == SET_RADIO_ARD ||
+            setup->bRequest == SET_RADIO_ARC ||
+            setup->bRequest == ACK_ENABLE ||
+            setup->bRequest == SET_CONT_CARRIER ||
+            setup->bRequest == SET_MODE ) {
+            
+            LOG_DBG("Queuing command %d", setup->bRequest);
+
+
+            command.type = command_setup;
+            uint32_t length = *len;
+            memcpy(&command.setup.setup_packet, setup, sizeof(struct usb_setup_packet));
+            if (length > sizeof(command.setup.data)) {
+                length = sizeof(command.setup.data);
             }
-			// If 250K is selected, packets will be ignored (ie. virtually not acked)
-			state.datarate = datarate;
-		} else if (setup->bRequest == SET_RADIO_POWER && setup->wLength == 0) {
-			LOG_DBG("Setting radio power %d", setup->wValue);
-            uint8_t power = MIN(setup->wValue, 3);
-            uint8_t fem_power = power_mapping[power];
-            fem_set_power(fem_power);
-		} else if (setup->bRequest == SET_RADIO_ARD && setup->wLength == 0) {
-			LOG_DBG("Setting radio ARD %d", setup->wValue);
-		} else if (setup->bRequest == SET_RADIO_ARC && setup->wLength == 0) {
-			LOG_DBG("Setting radio ARC %d", setup->wValue & 0x0f);
-            esb_set_arc(setup->wValue & 0x0f);
-		} else if (setup->bRequest == ACK_ENABLE && setup->wLength == 0) {
-            bool enabled = setup->wValue != 0;
-			LOG_DBG("Setting radio ACK Enable %s", enabled?"true":"false");
-            state.ack_enabled = enabled;
-            esb_set_ack_enabled(enabled);
-		} else if (setup->bRequest == SET_CONT_CARRIER && setup->wLength == 0) {
-			LOG_DBG("Setting radio Continious carrier %s", setup->wValue?"true":"false");
-            bool enable = setup->wValue != 0;
-            esb_set_continuous_carrier(enable);
-        } else if (setup->bRequest == CHANNEL_SCANN && usb_reqtype_is_to_device(setup)) {
+            if (usb_reqtype_is_to_device(setup) && length > 0)
+            {
+                memcpy(command.setup.data, *data, length);
+                command.setup.length = length;
+            }
+            k_msgq_put(&command_queue, &command, K_FOREVER);
+        } 
+        else if (setup->bRequest == CHANNEL_SCANN && usb_reqtype_is_to_device(setup)) {
+            k_mutex_lock(&usb_radio_mutex, K_FOREVER);
             uint8_t start = setup->wValue;
             uint8_t stop = setup->wIndex;
-            fw_scan(start, stop, *data,setup->wLength);
+            fw_scan(start, stop, *data, setup->wLength);
+            k_mutex_unlock(&usb_radio_mutex);
         } else if (setup->bRequest == CHANNEL_SCANN && usb_reqtype_is_to_host(setup)) {
             *data = state.scan_result;
             *len = MIN(state.scan_result_length, setup->wLength);
-		} else if (setup->bRequest == SET_MODE && setup->wLength == 0) {
-			LOG_DBG("Setting radio Mode %d", setup->wValue);
-        } else if (setup->bRequest == RESET_TO_BOOTLOADER && setup->wLength == 0) {
+        }
+        else if (setup->bRequest == RESET_TO_BOOTLOADER) {
+            LOG_DBG("Vendor request: RESET_TO_BOOTLOADER");
             system_reset_to_uf2();
-		} else {
-			return -ENOTSUP;
-		}
+        } else {
+            return -ENOTSUP;
+
+        }
+        
+		return 0;
     }
 
 	return 0;
@@ -249,64 +272,70 @@ static void usb_thread(void *, void *, void *) {
     while(1) {
         k_msgq_get(&command_queue, &command, K_FOREVER);
 
-        // If we are not receiving ack (ie. broadcast) and the received data is > 32 bytes,
-        // this means that the buffer actually contains 2 packets to send
-        if (!state.ack_enabled && command.length > 32) {
-            // Send the first one right away
-            memcpy(packet.data, command.payload, command.length/2);
-            packet.length = command.length/2;
-            esb_send_packet(&packet, &ack, &rssi, &arc_counter);
+        k_mutex_lock(&usb_radio_mutex, K_FOREVER);
+        if (command.type == command_data) {
+            // If we are not receiving ack (ie. broadcast) and the received data is > 32 bytes,
+            // this means that the buffer actually contains 2 packets to send
+            if (!state.ack_enabled && command.data.length > 32) {
+                // Send the first one right away
+                memcpy(packet.data, command.data.payload, command.data.length/2);
+                packet.length = command.data.length/2;
+                esb_send_packet(&packet, &ack, &rssi, &arc_counter);
 
-            // And prepare the second one to be send by the normal execution flow
-            memcpy(packet.data, &command.payload[command.length/2], command.length/2);
-            packet.length = command.length/2;
-        } else {
-            // Otherwise, cap to 32 bytes and prepare the unicast packets
-            if (command.length > 32) {
-                command.length = 32;
+                // And prepare the second one to be send by the normal execution flow
+                memcpy(packet.data, &command.data.payload[command.data.length/2], command.data.length/2);
+                packet.length = command.data.length/2;
+            } else {
+                // Otherwise, cap to 32 bytes and prepare the unicast packets
+                if (command.data.length > 32) {
+                    command.data.length = 32;
+                }
+                memcpy(packet.data, command.data.payload, command.data.length);
+                packet.length = command.data.length;
             }
-            memcpy(packet.data, command.payload, command.length);
-            packet.length = command.length;
-        }
-        
-        if (state.datarate != 0 && state.channel <= 100) {
-            bool acked = esb_send_packet(&packet, &ack, &rssi, &arc_counter);
-            if (ack.length > 32) {
-                LOG_DBG("Got an ack of size %d!", ack.length);
-            }
-            if (!state.ack_enabled) {
-                led_pulse_green(K_MSEC(50));
-            } else if (acked && ack.length <= 32) {
-                static char usb_answer[33];
-                usb_answer[0] = (arc_counter & 0x0f) << 4 | (rssi < 64)<<1 | 1;
-                memcpy(&usb_answer[1], ack.data, ack.length);
             
-                if (usb_write(CRAZYRADIO_IN_EP_ADDR, usb_answer, ack.length + 1, NULL)) {
-                    LOG_DBG("ep 0x%x", CRAZYRADIO_IN_EP_ADDR);
+            if (state.datarate != 0 && state.channel <= 100) {
+                bool acked = esb_send_packet(&packet, &ack, &rssi, &arc_counter);
+                if (ack.length > 32) {
+                    LOG_DBG("Got an ack of size %d!", ack.length);
+                }
+                if (!state.ack_enabled) {
+                    led_pulse_green(K_MSEC(50));
+                } else if (acked && ack.length <= 32) {
+                    static char usb_answer[33];
+                    usb_answer[0] = (arc_counter & 0x0f) << 4 | (rssi < 64)<<1 | 1;
+                    memcpy(&usb_answer[1], ack.data, ack.length);
+                
+                    if (usb_write(CRAZYRADIO_IN_EP_ADDR, usb_answer, ack.length + 1, NULL)) {
+                        LOG_DBG("ep 0x%x", CRAZYRADIO_IN_EP_ADDR);
+                    }
+
+                    led_pulse_green(K_MSEC(50));
+                } else {
+                    char no_ack_answer[1] = {0};
+            
+                    if (usb_write(CRAZYRADIO_IN_EP_ADDR, no_ack_answer, 1, NULL)) {
+                        LOG_DBG("ep 0x%x", CRAZYRADIO_IN_EP_ADDR);
+                    }
+
+                    led_pulse_red(K_MSEC(50));
                 }
 
-                led_pulse_green(K_MSEC(50));
             } else {
+                LOG_DBG("Not sending, radio settings not handled!");
                 char no_ack_answer[1] = {0};
-        
+            
                 if (usb_write(CRAZYRADIO_IN_EP_ADDR, no_ack_answer, 1, NULL)) {
                     LOG_DBG("ep 0x%x", CRAZYRADIO_IN_EP_ADDR);
                 }
 
                 led_pulse_red(K_MSEC(50));
             }
-
-        } else {
-            LOG_DBG("Not sending, radio settings not handled!");
-            char no_ack_answer[1] = {0};
-        
-            if (usb_write(CRAZYRADIO_IN_EP_ADDR, no_ack_answer, 1, NULL)) {
-                LOG_DBG("ep 0x%x", CRAZYRADIO_IN_EP_ADDR);
-            }
-
-            led_pulse_red(K_MSEC(50));
+        } else if (command.type == command_setup) {
+            LOG_DBG("Handling setup command %d", command.setup.setup_packet.bRequest);
+            handle_vendor_command(&command.setup);
         }
-
+        k_mutex_unlock(&usb_radio_mutex);
     }
 }
 
@@ -340,5 +369,52 @@ static void fw_scan(uint8_t start, uint8_t stop, char* data, int data_length) {
         } else {
             led_pulse_red(K_MSEC(50));
         }
+    }
+}
+
+static void handle_vendor_command(struct setup_command* setup) {
+    if (setup->setup_packet.bRequest == SET_RADIO_CHANNEL && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio channel %d", setup->setup_packet.wValue);
+        uint16_t channel = setup->setup_packet.wValue;
+        if (channel <= 100) {
+            esb_set_channel(channel);
+        }
+        // If channel >100 used, packets will be ignored (ie. virtually not acked)
+        state.channel = channel;
+    } else if (setup->setup_packet.bRequest == SET_RADIO_ADDRESS && setup->setup_packet.wLength == 5) {
+        LOG_DBG("Setting radio address %02x%02x%02x%02x%02x", (unsigned int)(setup->data)[0], (unsigned int)(setup->data)[1], (unsigned int)(setup->data)[2], (unsigned int)(setup->data)[3], (unsigned int)(setup->data)[4]);
+        esb_set_address(setup->data);
+    } else if (setup->setup_packet.bRequest == SET_DATA_RATE && setup->setup_packet.wLength == 0 && setup->setup_packet.wValue < 3) {
+        char* datarates[] = {"250K", "1M", "2M"};
+        LOG_DBG("Setting radio datarate to %s", datarates[setup->setup_packet.wValue]);
+        uint32_t datarate = setup->setup_packet.wValue;
+        if (datarate == 1) {
+            esb_set_bitrate(radioBitrate1M);
+        } else if (datarate == 2) {
+            esb_set_bitrate(radioBitrate2M);
+        }
+        // If 250K is selected, packets will be ignored (ie. virtually not acked)
+        state.datarate = datarate;
+    } else if (setup->setup_packet.bRequest == SET_RADIO_POWER && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio power %d", setup->setup_packet.wValue);
+        uint8_t power = MIN(setup->setup_packet.wValue, 3);
+        uint8_t fem_power = power_mapping[power];
+        fem_set_power(fem_power);
+    } else if (setup->setup_packet.bRequest == SET_RADIO_ARD && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio ARD %d", setup->setup_packet.wValue);
+    } else if (setup->setup_packet.bRequest == SET_RADIO_ARC && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio ARC %d", setup->setup_packet.wValue & 0x0f);
+        esb_set_arc(setup->setup_packet.wValue & 0x0f);
+    } else if (setup->setup_packet.bRequest == ACK_ENABLE && setup->setup_packet.wLength == 0) {
+        bool enabled = setup->setup_packet.wValue != 0;
+        LOG_DBG("Setting radio ACK Enable %s", enabled?"true":"false");
+        state.ack_enabled = enabled;
+        esb_set_ack_enabled(enabled);
+    } else if (setup->setup_packet.bRequest == SET_CONT_CARRIER && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio Continious carrier %s", setup->setup_packet.wValue?"true":"false");
+        bool enable = setup->setup_packet.wValue != 0;
+        esb_set_continuous_carrier(enable);
+    } else if (setup->setup_packet.bRequest == SET_MODE && setup->setup_packet.wLength == 0) {
+        LOG_DBG("Setting radio Mode %d", setup->setup_packet.wValue);
     }
 }

